File: android/src/main/kotlin/com/deepak/device_utils/DeviceUtilsPlugin.kt
package com.deepak.device_utils

import android.content.*
import android.os.Build
import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel

/** DeviceUtilsPlugin */
class DeviceUtilsPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "device_utils")
        channel.setMethodCallHandler(this)

        BatteryEventChannel.setupStatus(flutterPluginBinding.binaryMessenger, context)
        BatteryEventChannel.setupLevel(flutterPluginBinding.binaryMessenger, context)
        NetworkEventChannel.setupStatus(flutterPluginBinding.binaryMessenger, context)

    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "battery-level") {
            val batteryLevel = Utils.getBatteryLevel(context)
            if (batteryLevel != -1) {
                result.success(batteryLevel)
            } else {
                result.error("UNAVAILABLE", "Battery level not available.", null)
            }
        } else if (call.method == "network-is-connected") {
            result.success(Utils.isConnected(context))
        } else if (call.method == "storage-info") {
            var temp = Utils.storageInfo(context);
            result.success(temp)
        } else if (call.method == "device-info") {
            var temp = Utils.getDeviceInfo(context);
            result.success(temp)
        } else {
            result.notImplemented()
        }
    }


    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        BatteryEventChannel.close(context)
        NetworkEventChannel.close(context)
    }
}

File: ios/Classes/DeviceUtilsPlugin.swift
import Flutter
import UIKit

public class DeviceUtilsPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "device_utils", binaryMessenger: registrar.messenger())
    let instance = DeviceUtilsPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)


// Battery level EventChannel
        let levelChannel = FlutterEventChannel(name: "device_utils/battery-level-event", binaryMessenger: registrar.messenger())
        let levelHandler = BatteryLevelStreamHandler()
        levelChannel.setStreamHandler(levelHandler)
//         instance.batteryLevelStreamHandler = levelHandler


// Battery state EventChannel
        let stateChannel = FlutterEventChannel(name: "device_utils/battery-status-event", binaryMessenger: registrar.messenger())
        let stateHandler = BatteryStateStreamHandler()
        stateChannel.setStreamHandler(stateHandler)
//         instance.batteryStateStreamHandler = stateHandler

        let networkChannel = FlutterEventChannel(name: "device_utils/network-status-event", binaryMessenger: registrar.messenger())
        let networkHandler = NetworkStatusStreamHandler()
        networkChannel.setStreamHandler(networkHandler)
//         instance.batteryStateStreamHandler = stateHandler

  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "battery-level":
       result(Utils.getBatteryLevel())
    case "network-is-connected":
       result(Utils.isConnected())
    case "storage-info":
      result(Utils.getDeviceResources())
    case "device-info":
      result(Utils.getDeviceInfo())
    case "device-vibrate":
      Utils.vibrateDevice()
      result(true)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
